homeassistant/components/zwave_js/services.py

```diff
@@ -0,0 +1,112 @@
+"""Methods and classes related to executing Z-Wave commands and publishing these to hass."""
+
+import logging
+from typing import Dict, List, Union
+
+import voluptuous as vol
+from zwave_js_server.model.node import Node as ZwaveNode
+from zwave_js_server.util.node import async_set_config_parameter
+
+from homeassistant.const import ATTR_DEVICE_ID, ATTR_ENTITY_ID
+from homeassistant.core import HomeAssistant, ServiceCall, callback
+import homeassistant.helpers.config_validation as cv
+
+from . import const
+from .helpers import async_get_node_from_device_id, async_get_node_from_entity_id
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def parameter_name_does_not_need_bitmask(
+    val: Dict[str, Union[int, str]]
+) -> Dict[str, Union[int, str]]:
+    """Validate that if a parameter name is provided, bitmask is not as well."""
+    if isinstance(val[const.ATTR_CONFIG_PARAMETER], str) and (
+        val.get(const.ATTR_CONFIG_PARAMETER_BITMASK)
+    ):
+        raise vol.Invalid(
+            "Don't include a bitmask when a parameter name is specified",
+            path=[const.ATTR_CONFIG_PARAMETER, const.ATTR_CONFIG_PARAMETER_BITMASK],
+        )
+    return val
+
+
+# Validates that a bitmask is provided in hex form and converts it to decimal
+# int equivalent since that's what the library uses
+BITMASK_SCHEMA = vol.All(
+    cv.string, vol.Lower, vol.Match(r"^(0x)?[0-9a-f]+$"), lambda value: int(value, 16)
+)
+
+
+class ZWaveServices:
+    """Class that holds our services (Zwave Commands) that should be published to hass."""
+
+    def __init__(self, hass: HomeAssistant):
+        """Initialize with hass object."""
+        self._hass = hass
+
+    @callback
+    def async_register(self) -> None:
+        """Register all our services."""
+        self._hass.services.async_register(
+            const.DOMAIN,
+            const.SERVICE_SET_CONFIG_PARAMETER,
+            self.async_set_config_parameter,
+            schema=vol.All(
+                {
+                    vol.Optional(ATTR_DEVICE_ID, "id"): vol.All(
```

Comment:
The second parameter to `vol.Optional` is `msg`. Is that intended?
https://alecthomas.github.io/voluptuous/docs/_build/html/voluptuous.html#voluptuous.schema_builder.Optional

Reply:
this used to be Exclusive with ATTR_ENTITY_ID. Not sure when that changed, which makes more sense? To me you should only be providing one or the other, and this places nicely with the new target UI since I believe it will just translate entities, devices, and areas to a list of entities

Reply:
Looks like it was changed in this commit: https://github.com/home-assistant/core/pull/46673/commits/6f30b58267e92fe9f721463d57956093a9c89dba

Reply:
It seems we could support both at the same time, since we'll translate them to nodes. Do we need to make sure we don't target the same node more than once?

Reply:
good idea